Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateServiceImpl.java

```diff
@@ -0,0 +1,163 @@
+import com.serotonin.m2m2.vo.permission.PermissionException;
+
+import au.com.bytecode.opencsv.CSVReader;
+
+@Service
+public class ConfigurationTemplateServiceImpl implements ConfigurationTemplateService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationTemplateServiceImpl.class);
+
+    private final FileStoreService fileStoreService;
+
+    private final PermissionService permissionService;
+
+    public ConfigurationTemplateServiceImpl(FileStoreService fileStoreService, PermissionService permissionService) {
+        this.fileStoreService = fileStoreService;
+        this.permissionService = permissionService;
+    }
+
+    public List<Map<String, Object>> generateConfig(String fileStore, String filePath, List<Map<String, String>> keys) throws IOException,
+            PermissionException {
+        permissionService.ensureAdminRole(Common.getUser());
+
+        Path path = fileStoreService.getPathForRead(fileStore, filePath);
+        try (Reader in = new InputStreamReader(Files.newInputStream(path), Charset.defaultCharset())) {
+            CSVReader csvReader = new CSVReader(in);
+            List<Map<String, Object>> readFileStructure = readCSV(csvReader);
+
+            List<Map<String, Object>> result = groupByKeys(readFileStructure, keys);
+            return result;
+        }
+    }
+
+    public String generateMangoConfigurationJson(String fileStore, String filePath, String template, List<Map<String, String>> keys) throws IOException,
+            PermissionException {
+        MustacheFactory mf = new DefaultMustacheFactory();
+        List<Map<String, Object>> sites = generateConfig(fileStore, filePath, keys);
+
+        ClassPathResource cp = new ClassPathResource(template);
+        Mustache m = mf.compile(cp.getPath());
+        StringWriter writer = new StringWriter();
+        m.execute(writer, sites).flush();
+        StringBuilder result = new StringBuilder(writer.toString());
+
+        //TODO: Fix this trailing comma situation
+        int trailingComma = result.lastIndexOf(",");
+        result.delete(trailingComma, trailingComma+1);
+        return result.toString();
+    }
+
+    /**
+     * returns a group from the list of map<String, Object> based on the provided keys.
+     *
+     * @param array - List of data from the data source file.
+     * @param key - map of keys used for structuring data for the configuration file.
+     * @return grouped data based on the provided keys in a hierarchical structure.
+     */
+    private List<Map<String, Object>> groupByKey(List<Map<String, Object>> array, Map<String, String> key) {
+        Map<Object, Map<String, Object>> map = new HashMap<>();
+        for (Map<String, Object> item : array) {
+            Object keyValue = item.get(key.get("groupKey"));
+            if (!map.containsKey(keyValue)) {
+                Map<String, Object> group = new HashMap<>();
+                group.put(key.get("groupKey"), keyValue);
```

Comment:
I don't think you can hard code the keys of the key map.  @jazdw can you elaborate on the purpose of the `key` map?

Reply:
I thought this key name was some sort of standard for the groupKey element and the childrenKey element.